README.md
# socket-signal

[![Build Status](https://travis-ci.com/geut/socket-signal.svg?branch=master)](https://travis-ci.com/geut/socket-signal)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
[![standard-readme compliant](https://img.shields.io/badge/readme%20style-standard-brightgreen.svg?style=flat-square)](https://github.com/RichardLitt/standard-readme)

> Signal abstraction to create WebRTC connections through sockets.

This module provides a common interface (for the client and server) to build your own signal.

If you want to use a `ready to use` implementation check: [socket-signal-websocket](https://github.com/geut/socket-signal-websocket).

## <a name="install"></a> Install

```
$ npm install socket-signal
```

## <a name="usage"></a> Usage

### Client

The client interface provides almost everything that you need to start a basic signal client (most of the times you won't need to add more features).

```javascript
const { SocketSignalClient } = require('socket-signal')

class YourClient extends SocketSignalClient {
  async _open() {
    await super._open()
  }

  async _close() {
    await super._close()
  }

  async _onOffer(data) {
    // data.offer
  }

  async _onAnswer(data) {
    // data.answer
  }

  async _onIncomingPeer(peer) {
    if (validPeer(peer)) return
    throw new Error('invalid peer')
  }
}

const client = new YourClient(socket, opts)

;(async () => {
  // open the client
  await client.open()

  // join the swarm for the given topic and get the current peers for that topic
  const peersAvailable = await client.join(topic)

  // request a connection to a specific peer
  const remotePeer = client.connect(peersAvailable[0], topic)

  // optional, use the signal to listen for incoming media streams
  remotePeer.subscribeMediaStream = true

  try {
    // wait until the connection is established
    await remotePeer.ready()
    console.log(remotePeer.stream)
    // SimplePeer connected
  } catch(err) {
    // SimplePeer rejected
  }
})()
```

### Server

```javascript
const { SocketSignalServer } = require('socket-signal')

class YourServer extends SocketSignalServer {
  async _onDisconnect (rpc) {
    // onDisconnect a peer socket
  }

  async _onJoin (rpc, msg) {
    // msg.id, msg.topic
  }

  async _onLeave (rpc, msg) {
    // msg.id, msg.topic
  }

  async _onLookup (rpc, msg) {
    // msg.topic
  }

  /**
   * request signal offer
   */
  async _onOffer (rpc, msg) {
    // msg.remoteId, msg.topic, msg.data
  }

  /**
   * event signal
   *
   * this event is emitted when there is a signal candidate
   */
  async _onSignal (rpc, msg) {
    // msg.remoteId, msg.topic, msg.data
  }
}
```

> If you want a server with a minimal implementation to handle peer connections you can use [SocketSignalServerMap](lib/server-map.js).

## API

### Client

#### `const client = new Client(options)`

Creates a new client instance.

Options include:

- `id: Buffer`: ID of 32 bytes.
- `requestTimeout: 15 * 1000`: How long to wait for peer requests.
- `queueConcurrency: 4`: How many incoming connections in concurrent can handle
- `metadata: Object`: Metadata to share across network.
- `simplePeer: Object`: SimplePeer options.

> IMPORTANT: Every `id` and `topic` must be a `Buffer of 32 bytes`.

#### `client.open() => Promise`

Open the client.

#### `client.close() => Promise`

Close the client.

#### `client.join(topic: Buffer) => Promise<Array<Buffer>>`

Join the swarm for the given topic and do a `lookup` of peers available for that topic.

#### `client.leave(topic: Buffer) => Promise`

Leave the swarm for the given topic.

> IMPORTANT: This will not close the current peers for that topic you should call `closeConnectionsByTopic(topic)`

#### `client.closeConnectionsByTopic(topic: Buffer) => Promise`

Close all the connections referenced by a topic.

#### `client.connect(peerId: Buffer, topic?: Buffer, options?: Object) => Peer`

Creates a `request` connection for a specific peer and topic. `topic` is optional.

`options` include:

- `metadata: Object`: Metadata to share with the other peer.
- `simplePeer: Object`: Specific SimplePeer options for this connection.

`Peer` is an object with:

- `id: Buffer`: ID of the peer.
- `sessionId: Buffer`: Unique ID for this connection.
- `topic: Buffer`: Topic related.
- `localMetadata: Object`: Your metadata shared with the peer.
- `metadata: Object`: The remote metadata, belongs to the peer connected to.
- `stream: SimplePeer`: The SimplePeer internal stream.
- `subscribeMediaStream: boolean`: Set in `true` if you want to use the signal to listen for incoming media streams. Default: `false`.

#### `peer.ready() => Promise`

Wait for the connection to be established.

#### `peer.addStream(mediaStream) => Peer`

Add a media stream.

## <a name="issues"></a> Issues

:bug: If you found an issue we encourage you to report it on [github](https://github.com/geut/socket-signal/issues). Please specify your OS and the actions to reproduce it.

## <a name="contribute"></a> Contributing

:busts_in_silhouette: Ideas and contributions to the project are welcome. You must follow this [guideline](https://github.com/geut/socket-signal/blob/master/CONTRIBUTING.md).

## License

MIT © A [**GEUT**](http://geutstudio.com/) project


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## 9.2.0 - 2021-02-04 [YANKED]

## 9.1.0 - 2021-01-27 [YANKED]

## 9.0.0 - 2020-08-21 [YANKED]

## 8.1.0 - 2020-08-19 [YANKED]

## 8.0.1 - 2020-08-19 [YANKED]

## 8.0.0 - 2020-08-12 [YANKED]

## 7.0.1 - 2020-07-31 [YANKED]

## 7.0.0 - 2020-07-28 [YANKED]

## 6.0.1 - 2020-07-26 [YANKED]

## 6.0.0 - 2020-07-26 [YANKED]

## 5.1.0 - 2020-07-21 [YANKED]

## 5.0.0 - 2020-05-26 [YANKED]

## 4.0.1 - 2020-04-22 [YANKED]

## 4.0.0 - 2020-04-22 [YANKED]

## 3.0.2 - 2020-04-22 [YANKED]

## 3.0.1 - 2020-04-22 [YANKED]

## 3.0.0 - 2020-04-21 [YANKED]

## 2.0.0 - 2020-04-20 [YANKED]

## 1.0.12 - 2020-04-02 [YANKED]

## 1.0.11 - 2020-03-27 [YANKED]

## 1.0.10 - 2020-03-17 [YANKED]

## 1.0.9 - 2020-03-16 [YANKED]

## 1.0.8 - 2020-03-16 [YANKED]

## 1.0.7 - 2020-03-16 [YANKED]

## 1.0.6 - 2020-03-16 [YANKED]

## 1.0.5 - 2020-03-16 [YANKED]

## 1.0.4 - 2020-03-16 [YANKED]

## 1.0.3 - 2020-03-16 [YANKED]

## 1.0.2 - 2020-03-16 [YANKED]

## 1.0.1 - 2020-03-15 [YANKED]

## 1.0.0 - 2020-03-15 [YANKED]
[Unreleased]: https://github.com/geut/socket-signal/compare/v9.2.0...HEAD


lib/client.js
const crypto = require('crypto')

const assert = require('nanocustomassert')
const fastq = require('fastq')
const pEvent = require('p-event')
const nanomessagerpc = require('nanomessage-rpc')
const { NanoresourcePromise } = require('nanoresource-promise/emitter')

const Peer = require('./peer')

const kConnectionsQueue = Symbol('socketsignal.connectionsqueue')
const kPeers = Symbol('socketsignal.peers')
const kDefineActions = Symbol('socketsignal.defineactions')
const kDefineEvents = Symbol('socketsignal.defineevents')
const kAddPeer = Symbol('socketsignal.addpeer')
const kCreatePeer = Symbol('socketsignal.createpeer')
const kOnSignal = Symbol('socketsignal.onsignal')

function worker ({ peer, data }, done) {
  this.open()
    .then(() => {
      this.emit('peer-connecting', peer)
      return peer.open(data)
    })
    .then(() => done())
    .catch(err => done(err))
}

class SocketSignalClient extends NanoresourcePromise {
  /**
   * @constructor
   * @param {DuplexStream} socket
   * @param {Object} opts
   * @param {Buffer} opts.id Id of 32 bytes
   * @param {number} opts.requestTimeout How long to wait for peer requests
   * @param {number} opts.queueConcurrency How many incoming connections in concurrent can handle
   * @param {Object} opts.metadata Metadata to share across network
   * @param {Object} opts.simplePeer SimplePeer options
   */
  constructor (socket, opts = {}) {
    super()

    const {
      id = crypto.randomBytes(32),
      requestTimeout = 15 * 1000,
      queueConcurrency = 4,
      simplePeer = {},
      metadata
    } = opts

    assert(!metadata || typeof metadata === 'object', 'metadata must be an object')

    this.socket = socket
    this.rpc = nanomessagerpc({ timeout: requestTimeout, ...nanomessagerpc.useSocket(socket) })
    this.id = id
    this.metadata = metadata
    this.simplePeer = simplePeer
    this.requestTimeout = requestTimeout

    this[kPeers] = new Map()
    this[kConnectionsQueue] = fastq(this, worker, queueConcurrency)

    // rpc
    this[kDefineActions]()
    this[kDefineEvents]()
  }

  /**
   * Peers
   *
   * @type {Array<Peer>}
   */
  get peers () {
    return Array.from(this[kPeers].values())
  }

  /**
   * Peers connected
   *
   * @type {Array<Peer>}
   */
  get peersConnected () {
    return this.peers.filter(p => p.connected)
  }

  /**
   * Peers incoming and connecting
   *
   * @type {Array<Peer>}
   */
  get peersConnecting () {
    return this.peers.filter(p => !p.connected)
  }

  /**
   * Get peers by the topic
   *
   * @param {Buffer} topic
   * @returns {Array<Peer>}
   */
  getPeersByTopic (topic) {
    assert(Buffer.isBuffer(topic), 'topic is required')

    return this.peers.filter(peer => peer.topic.equals(topic))
  }

  /**
   * Join to the network by a topic
   *
   * @param {Buffer} topic
   * @returns {Promise<Array<Peer>>}
   */
  async join (topic) {
    assert(Buffer.isBuffer(topic) && topic.length === 32, 'topic must be a Buffer of 32 bytes')

    await this.open()
    const peers = await this.rpc.call('join', this._buildMessage({ topic }))
    this.emit('join', topic, peers)
    return peers
  }

  /**
   * Leave a topic from the network
   *
   * IMPORTANT: This will not close the current peers for that topic
   * you should call closeConnectionsByTopic(topic)
   *
   * @param {Buffer} topic
   * @returns {Promise}
   */
  async leave (topic) {
    assert(!topic || (Buffer.isBuffer(topic) && topic.length === 32), 'topic must be a Buffer of 32 bytes')

    await this.open()
    await this.rpc.call('leave', this._buildMessage({ topic }))
    this.emit('leave', topic)
  }

  /**
   * Calls a new lookup from the network
   *
   * @param {Buffer} topic
   * @returns {Promise<Array<Peer>>}
   */
  async lookup (topic) {
    assert(Buffer.isBuffer(topic) && topic.length === 32, 'topic must be a Buffer of 32 bytes')

    await this.open()
    const peers = await this.rpc.call('lookup', this._buildMessage({ topic }))
    this.emit('lookup', topic, peers)
    return peers
  }

  /**
   * Connects to a peer by their id and topic
   *
   * IMPORTANT: This will not returns a connected peer
   * you should wait for the connection by peer.ready()
   *
   * @param {Buffer} id
   * @param {Buffer} topic
   * @param {(Object|undefined)} opts
   * @param {Object} opts.metadata
   * @param {Object} opts.simplePeer
   * @returns {Peer}
   */
  connect (id, topic, opts = {}) {
    assert(Buffer.isBuffer(id) && id.length === 32, 'id must be a Buffer of 32 bytes')
    assert(!topic || (Buffer.isBuffer(topic) && topic.length === 32), 'topic must be a Buffer of 32 bytes')

    const { metadata: localMetadata, simplePeer = {} } = opts

    const peer = this[kCreatePeer]({ initiator: true, sessionId: crypto.randomBytes(32), id, topic, localMetadata, simplePeer })

    this[kAddPeer](peer)

    return peer
  }

  /**
   * Async handler for incoming peers, peers that you don't get it from .connect(id, topic)
   *
   * This is the right place to define rules to accept or reject connections.
   *
   * @param {(peer: Peer) => (Promise|Error)} handler
   */
  onIncomingPeer (handler) {
    this._onIncomingPeer = handler
    return this
  }

  /**
   * Async handler for incoming offer.
   *
   * @param {(data: { id, topic, metadata }) => (Promise|Error)} handler
   */
  onOffer (handler) {
    this._onOffer = handler
    return this
  }

  /**
   * Async handler for incoming answer.
   *
   * @param {(data: { id, topic, metadata }) => (Promise|Error)} handler
   */
  onAnswer (handler) {
    this._onAnswer = handler
    return this
  }

  /**
   * Close connections by topic
   *
   * @param {Buffer} topic
   * @returns {Promise}
   */
  closeConnectionsByTopic (topic) {
    return Promise.all(this.getPeersByTopic(topic).map(peer => new Promise(resolve => peer.close(() => resolve()))))
  }

  /**
   * @abstract
   */
  async _onIncomingPeer () {}

  /**
   * @abstract
   */
  async _onOffer (data) {}

  /**
   * @abstract
   */
  async _onAnswer (data) {}

  async _open () {
    await this.rpc.open()
  }

  async _close () {
    this[kConnectionsQueue].kill()
    await this.rpc.close()
    await Promise.all(this.peers.map(peer => new Promise(resolve => peer.close(() => resolve()))))
  }

  _buildMessage (data) {
    const { localMetadata = {}, ...msg } = data
    let metadata = this.metadata || {}
    if (localMetadata) {
      metadata = { ...metadata, ...localMetadata }
    }
    return { ...msg, id: this.id, metadata }
  }

  /**
   * @private
   */
  [kDefineActions] () {
    this.rpc.actions({
      offer: async (message) => {
        const result = await this._onOffer(message)

        assert(!result || typeof result === 'object')

        const peer = this[kCreatePeer]({
          initiator: false,
          sessionId: message.sessionId,
          id: message.id,
          topic: message.topic,
          metadata: message.metadata,
          localMetadata: result && result.metadata,
          simplePeer: result && result.simplePeer
        })

        this[kAddPeer](peer, message.data)

        return pEvent(peer, 'answer', {
          rejectionEvents: ['error', 'closed']
        })
      }
    })
  }

  /**
   * @private
   */
  [kDefineEvents] () {
    this.rpc.on('signal', (message) => {
      const { sessionId, data = [] } = message
      const peer = this[kPeers].get(sessionId.toString('hex'))
      if (!peer || peer.destroyed) return
      data.forEach(signal => peer.stream.signal(signal))
    })
  }

  /**
   * @private
   */
  [kAddPeer] (peer, data) {
    const sessionId = peer.sessionId.toString('hex')

    this[kPeers].set(sessionId, peer)
    peer.once('close', () => {
      this[kPeers].delete(sessionId)
    })
    peer.once('error', err => {
      this.emit('peer-error', err, peer)
    })

    this[kConnectionsQueue].push({ peer, data }, (err) => {
      if (err || this.closing || this.closed) return process.nextTick(() => peer.destroy(err))
      this.emit('peer-connected', peer)
    })

    // peer queue
    this.emit('peer-queue', peer)
  }

  /**
   * @private
   */
  [kCreatePeer] (opts) {
    opts = Object.assign({}, opts, {
      onSignal: (peer, batch) => this[kOnSignal](peer, batch),
      simplePeer: Object.assign({}, this.simplePeer, opts.simplePeer),
      timeout: this.requestTimeout
    })

    let peer
    if (typeof this.simplePeer === 'function') {
      peer = this.simplePeer(Peer, opts)
    }

    peer = new Peer(opts)

    return peer
  }

  /**
   * @private
   */
  async [kOnSignal] (peer, batch) {
    if (peer.destroyed || (peer.connected && !peer.subscribeMediaStream)) return

    const payload = () => this._buildMessage({
      remoteId: peer.id,
      topic: peer.topic,
      sessionId: peer.sessionId,
      data: batch,
      localMetadata: peer.localMetadata
    })

    if (!peer.connected) {
      const type = batch[0].type

      if (type === 'offer') {
        const response = await this.rpc.call('offer', payload())
        await this._onAnswer(response)
        response.data.forEach(signal => peer.stream.signal(signal))
        peer.metadata = response.metadata
        return
      }

      if (type === 'answer') {
        await this._onIncomingPeer(peer)
        return peer.emit('answer', payload())
      }
    }

    await this.rpc.emit('signal', payload())
  }
}

module.exports = SocketSignalClient
